src/borg/compress.pyx

```diff
@@ -433,6 +436,72 @@ class Auto(CompressorBase):
         raise NotImplementedError
 
 
+class ObfuscateSize(CompressorBase):
+    """
+    Meta-Compressor that obfuscates the compressed data size.
+    """
+
+    ID = b'\x04\x00'
+    name = 'obfuscate'
+
+    def __init__(self, level=None, compressor=None):
+        super().__init__()
+        self.compressor = compressor
+        if level == 1:
+            self.factor = 0.01
+            self.min_r = 0.0001
+        elif level == 2:
+            self.factor = 0.1
+            self.min_r = 0.0001
+        elif level == 3:
+            self.factor = 1.0
+            self.min_r = 0.0001
+        elif level == 4:
+            self.factor = 10.0
+            self.min_r = 0.0001
+        elif level == 5:
+            self.factor = 100.0
+            self.min_r = 0.0001
+        elif level == 6:
+            self.factor = 1000.0
+            self.min_r = 0.001
+        else:
+            # initialize for decompression
+            self.factor = None
+            self.min_r = None
+
+    def _obfuscate(self, compr_size):
+        # relative random reciprocal size variation
+        r = max(self.min_r, random.random())  # 0..1, but dont get too close to 0
+        f = self.factor / r
+        # effect for SPEC 1:
+        # f = 0.01 .. 0.1 for r in 1.0 .. 0.1 == in 90% of cases
+        # f = 0.1 .. 1.0 for r in 0.1 .. 0.01 == in 9% of cases
+        # f = 1.0 .. 10.0 for r in 0.01 .. 0.001 = in 0.9% of cases
+        # f = 10.0 .. 100.0 for r in 0.001 .. 0.0001 == in 0.09% of cases
+        addtl_size = int(compr_size * f)
+        addtl_size = min(MAX_DATA_SIZE - 1024 - compr_size, addtl_size)  # stay away from MAX_DATA_SIZE
+        return bytes(addtl_size)
```

Comment:
one issue with this:

at least for the smaller factors, due to the "relative" approach: small input often leads to a small output, also.

for big factors, this is less an issue at least for medium to big input, because that will get clipped to ~MAX_DATA_SIZE. very small inputs might still stick out i guess.

Reply:
implemented another obfuscator that just fills up to full blocks of a specific block size.

if arranged correctly, this can completely hide the compressed size, but can also waste a lot of space to do this.

Reply:
> at least for the smaller factors, due to the "relative" approach: small input often leads to a small output, also.

Adding another random increment that's uniformly random over a fixed fraction of the original size could help with that.

Reply:
Varying the compression algorithm and/or parameters at random for small files would add another level of difficulty to fingerprinting, if the architecture allows that.

Reply:
> implemented another obfuscator that just fills up to full blocks of a specific block size.
> 
> if arranged correctly, this can completely hide the compressed size, but can also waste a lot of space to do this.

I like the fact that this removes some size information completely, but I think it also allows an attacker in control of the repos of many users to determine which users are most willing to waste space for privacy, maybe that's undesirable.

Reply:
1. so it would e.g. add:

- 1kB * random() for input size 0..10kB
- 10kB * random() for input size 10..100kB
- 100kB * random() for inputs 100..1000kB
- 1000kB * random() for inputs 1000..8000kB

did you mean it like that?

2. architecturally difficult (with the current way used to implement this) and also it feels a bit strange/messy to vary the compression algos/params.

btw, compression usage also leaks a bit of information: stuff that is very good compressible gets very small. i noticed that when doing the unit tests. longer runs of zeros end up in tiny 20-30bytes of compressed data - and if one only adds random relative to that, it still is tiny. so, for the attacker: tiny -> runs of zeros (or other highly compressible data).

if one wants to avoid that, it might be better to not use compression.

3. that's a good point, yes, fixed block size somehow would "stick out" compared to no (or other) obfuscation.


Another idea: do it like the current "fill to full block" obfuscator, but compute the target blocksize:
- blocksize * random()  # might give tiny blocks
- blocksize * (0.5 + random())  # might "stick out" in statistics
- compute that once in `__init__`
- compute that for each compressor call

Reply:
ad 3: i solved that by not fully padding the final block, but randrange(0, full_pad) sized padding.

Reply:
hmm, need to improve that. if input sizes happen to be close to or match N*blocksize, padding size variation would be small or 0.

Reply:
ad 3, 2nd try: just adding a randomly sized padding now (up to the specified size). easier and better.